[TensorExpr] Fuser: rely on input types when checking whether a device is supported.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44139 [TensorExpr] Fuser: rely on input types when checking whether a device is supported.**

Also, make sure that we're checking that condition when we're starting a
new fusion group, not only when we merge a node into an existing fusion
group. Oh, and one more: add a test checking that we're rejecting graphs
with unspecified shapes.

Differential Revision: [D23507510](https://our.internmc.facebook.com/intern/diff/D23507510)

Commit:
[TensorExpr] Fuser: rely on input types when checking whether a device is supported.

Also, make sure that we're checking that condition when we're starting a
new fusion group, not only when we create a node into an existing fusion
group. Oh, and one more: add a test checking that we're rejecting graphs
with unspecified shapes.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: rely on input types when checking whether a device is supported."


Also, make sure that we're checking that condition when we're starting a
new fusion group, not only when we create a node into an existing fusion
group. Oh, and one more: add a test checking that we're rejecting graphs
with unspecified shapes.

Differential Revision: [D23507510](https://our.internmc.facebook.com/intern/diff/D23507510)

[ghstack-poisoned]

bertmaher reviewed (approved):
Oops.  Thanks for catching my bugs.

Commit:
Update on "[TensorExpr] Fuser: rely on input types when checking whether a device is supported."


Also, make sure that we're checking that condition when we're starting a
new fusion group, not only when we merge a node into an existing fusion
group. Oh, and one more: add a test checking that we're rejecting graphs
with unspecified shapes.

Differential Revision: [D23507510](https://our.internmc.facebook.com/intern/diff/D23507510)

[ghstack-poisoned]
